beacon-chain/rpc/prysm/v1alpha1/validator/duties_v2.go

```diff
@@ -0,0 +1,266 @@
+	}
+	meta.committeesAtSlot = helpers.SlotCommitteeCount(activeValidatorCount)
+
+	nextActiveValidatorCount, err := helpers.ActiveValidatorCount(ctx, s, reqEpoch+1)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get active validator count: %v", err)
+	}
+	meta.nextCommitteesAtSlot = helpers.SlotCommitteeCount(nextActiveValidatorCount)
+
+	meta.proposalSlots, err = helpers.ProposerAssignments(ctx, s, reqEpoch)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not compute proposer slots: %v", err)
+	}
+
+	return meta, nil
+}
+
+// buildValidatorDuty builds both current‑epoch and next‑epoch V2 duty objects
+// for a single validator index.
+func (vs *Server) buildValidatorDuty(
+	pubKey []byte,
+	idx primitives.ValidatorIndex,
+	s state.BeaconState,
+	reqEpoch primitives.Epoch,
+	meta *dutiesMetadata,
+) (*ethpb.DutiesV2Response_Duty, *ethpb.DutiesV2Response_Duty, error) {
+	assignment := &ethpb.DutiesV2Response_Duty{PublicKey: pubKey}
+	nextAssignment := &ethpb.DutiesV2Response_Duty{PublicKey: pubKey}
+
+	statusEnum := assignmentStatus(s, idx)
+
+	assignment.ValidatorIndex = idx
+	assignment.Status = statusEnum
+	// note: we only set the proposer slots for the current assignment and not the next epoch assignment
+	assignment.ProposerSlots = meta.proposalSlots[idx]
+	assignment.CommitteesAtSlot = meta.committeesAtSlot
+
+	nextAssignment.ValidatorIndex = idx
+	nextAssignment.Status = statusEnum
+
+	// Current epoch committee
+	if ca, ok := meta.assignments[idx]; ok {
+		populateCommitteeFields(assignment, ca)
+	}
+	// Next epoch committee
+	if ca, ok := meta.nextEpochAssignments[idx]; ok {
+		populateCommitteeFields(nextAssignment, ca)
+		nextAssignment.CommitteesAtSlot = meta.nextCommitteesAtSlot
+	}
+
+	// Sync committee flags
+	if coreTime.HigherEqualThanAltairVersionAndEpoch(s, reqEpoch) {
+		inSync, err := helpers.IsCurrentPeriodSyncCommittee(s, idx)
+		if err != nil {
+			return nil, nil, status.Errorf(codes.Internal, "Could not determine current epoch sync committee: %v", err)
+		}
+		assignment.IsSyncCommittee = inSync
+		nextAssignment.IsSyncCommittee = inSync
+		if inSync {
+			if err := core.RegisterSyncSubnetCurrentPeriodProto(s, reqEpoch, pubKey, statusEnum); err != nil {
+				return nil, nil, err
+			}
+		}
+
+		// Boundary check: determine if next period differs
+		nextEpoch := reqEpoch + 1
+		currentEpoch := coreTime.CurrentEpoch(s)
+		n := slots.SyncCommitteePeriod(nextEpoch)
+		c := slots.SyncCommitteePeriod(currentEpoch)
+		if n > c {
+			nextInSync, err := helpers.IsNextPeriodSyncCommittee(s, idx)
+			if err != nil {
+				return nil, nil, status.Errorf(codes.Internal, "Could not determine next epoch sync committee: %v", err)
+			}
+			nextAssignment.IsSyncCommittee = nextInSync
+			if nextInSync {
+				if err := core.RegisterSyncSubnetNextPeriodProto(s, reqEpoch, pubKey, statusEnum); err != nil {
+					return nil, nil, err
+				}
+			}
```

Comment:
Is this required in the critical path or can it be done in the background?

Reply:
This is something i'm not sure about